src/Microsoft.Extensions.ML/PredictionEnginePool.cs

```diff
@@ -100,7 +101,7 @@ public ITransformer GetModel()
             //Here we are in the world of named models where the model hasn't been built yet.
             var options = _predictionEngineOptions.Create(modelName);
             var pool = new PoolLoader<TData, TPrediction>(_serviceProvider, options);
-            _namedPools.Add(modelName, pool);
+            _namedPools.TryAdd(modelName, pool);
```

Comment:
⚠️  This implementation can end up using multiple pools for the same key. If you don't want that to happen, use `GetOrAdd` instead of `TryAdd`.

```suggestion
            pool = _namedPools.GetOrAdd(modelName, pool);
```

Reply:
- This implementation can end up using multiple pools for the same key
How can this happen? 
What's the difference between TryAdd and GetOrAdd?

---
In reply to: [357425395](https://github.com/dotnet/machinelearning/pull/4570#discussion_r357425395) [](ancestors = 357425395)

Reply:
> How can this happen?

Two threads call `TryAdd` at the same time with different pools. Only one will end up in `_namedPools`, yet each thread uses the pool it created itself.

> What's the difference between TryAdd and GetOrAdd?

In the presence of multiple callers, `TryAdd` will use the value from the first and ignore the others. `GetOrAdd` will atomically add the value or return the existing value.

Reply:
I see, thanks for explain!

---
In reply to: [357434533](https://github.com/dotnet/machinelearning/pull/4570#discussion_r357434533) [](ancestors = 357434533)